server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/builder/AbstractDocumentsFileManager.java

```diff
@@ -0,0 +1,232 @@
+    }
+
+    /**
+     * Read dimensions from file
+     */
+    protected long readDimensions(Long[] dimensions, RandomAccessInput input, long offset) throws IOException {
+        for (int i = 0; i < dimensions.length; i++) {
+            try {
+                dimensions[i] = input.readLong(offset);
+            } catch (Exception e) {
+                logger.error("Error reading dimension value at offset {} for dimension {}", offset, i);
+                throw e;
+            }
+            offset += Long.BYTES;
+        }
+        offset += StarTreeDocumentBitSetUtil.readBitSet(input, offset, dimensions, index -> null);
+        return offset;
+    }
+
+    /**
+     * Read star tree metrics from file
+     */
+    protected long readMetrics(RandomAccessInput input, long offset, int numMetrics, Object[] metrics, boolean shouldReadAggregatedDocs)
+        throws IOException {
+        for (int i = 0; i < numMetrics; i++) {
+            switch (metricAggregatorInfos.get(i).getValueAggregators().getAggregatedValueType()) {
+                case LONG:
+                    metrics[i] = input.readLong(offset);
+                    offset += Long.BYTES;
+                    break;
+                case DOUBLE:
+                    long val = input.readLong(offset);
+                    if (shouldReadAggregatedDocs) {
+                        metrics[i] = StarTreeNumericTypeConverters.sortableLongtoDouble(val);
+                    } else {
+                        metrics[i] = val;
+                    }
+                    offset += Long.BYTES;
+                    break;
+                default:
+                    throw new IllegalStateException("Unsupported metric type");
+            }
+        }
+        offset += StarTreeDocumentBitSetUtil.readBitSet(
+            input,
+            offset,
+            metrics,
+            index -> metricAggregatorInfos.get(index).getValueAggregators().getIdentityMetricValue()
+        );
+        return offset;
+    }
+
+    /**
+     * Write star tree document to file
+     */
+    public abstract void writeStarTreeDocument(StarTreeDocument starTreeDocument, boolean isAggregatedDoc) throws IOException;
+
+    /**
+     * Read star tree document from file based on doc id
+     */
+    public abstract StarTreeDocument readStarTreeDocument(int docId, boolean isMerge) throws IOException;
+
+    /**
+     * Read star document dimensions from file based on doc id
+     */
+    public abstract Long[] readDimensions(int docId) throws IOException;
+
+    /**
+     * Read dimension value for given doc id and dimension id
+     */
+    public abstract Long getDimensionValue(int docId, int dimensionId) throws IOException;
+
+    /**
+     * Delete the temporary files created
+     */
+    public void deleteFiles(boolean success) throws IOException {
+        if (success) {
+            for (String file : tmpDirectory.getCreatedFiles()) {
+                tmpDirectory.deleteFile(file);
```

Comment:
do we need error handling here for any possible failed deletions?

Reply:
No intentionally keeping it this way, in case of "success" , delete file should throw IOException.